nixos/tests/incus/container.nix

```diff
@@ -0,0 +1,113 @@
+      memorySize = 512;
+
+      incus.enable = true;
+    };
+  };
+
+  testScript = ''
+    def instance_is_up(_) -> bool:
+      status, _ = machine.execute("incus exec container --disable-stdin --force-interactive /run/current-system/sw/bin/true")
+      return status == 0
+
+    machine.wait_for_unit("incus.service")
+
+    # no preseed should mean no service
+    machine.fail("systemctl status incus-preseed.service")
+
+    machine.succeed("incus admin init --minimal")
+
+    machine.succeed(
+        "incus image import ${container-image-metadata}/*/*.tar.xz ${container-image-rootfs}/*/*.tar.xz --alias nixos"
+    )
+
+    with subtest("Container can be managed"):
+        machine.succeed("incus launch nixos container")
+        with machine.nested("Waiting for instance to start and be usable"):
+          retry(instance_is_up)
+        machine.succeed("echo true | incus exec container /run/current-system/sw/bin/bash -")
+        machine.succeed("incus delete -f container")
+
+    with subtest("Container is mounted with lxcfs inside"):
+        machine.succeed("incus launch nixos container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        ## ---------- ##
+        ## limits.cpu ##
+
+        machine.succeed("incus config set container limits.cpu 1")
+        machine.succeed("incus restart container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "1"
+            == machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        )
+
+        machine.succeed("incus config set container limits.cpu 2")
+        machine.succeed("incus restart container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "2"
+            == machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        )
+
+        ## ------------- ##
+        ## limits.memory ##
+
+        machine.succeed("incus config set container limits.memory 64MB")
+        machine.succeed("incus restart container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "MemTotal:          62500 kB"
+            == machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
+        )
+
+        machine.succeed("incus config set container limits.memory 128MB")
+        machine.succeed("incus restart container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "MemTotal:         125000 kB"
+            == machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
+        )
+
+        machine.succeed("incus delete -f container")
```

Comment:
If you end up making the CPU and memory section into their own subtests, just make sure this line doesn't accidentally end up as part of the memory test.

Reply:
I didn't. There's no need to create a whole separate container for these resource checks.

Reply:
I'm not sure we're talking about the same thing. With "subtest", I mean the python `with subtest("set limits.memory"):` log grouping feature.

Reply:
@adamcstephens Hope you don't mind me tagging you here, just in case you didn't see the last comment

Reply:
No, I don't mind at all. Maybe I wasn't clear, but I thought I understood your suggestion here.

I opted to not make the CPU and memory tests separate subtests, so it wasn't necessary to create and delete multiple containers. Or am I missing something else?

Reply:
I think the origin of my confusion is the "create and delete multiple containers" thing.

Unless I've misunderstood it completely, the subtest statement doesn't create a new test that requires you to spin up another VM, it just groups the different parts of a test under logical sections, in kind of the same way that you are grouping them with the comment titles. They even tell you which part failed if there's an error.

Please have a look at https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/test-driver/test_driver/driver.py#L88-L96

Reply:
Right, it wouldn’t necessarily require using multiple separate containers. Sorry for conflating the two.

Do you think the test as written is ok with my latest changes?

Reply:
Regarding the node-wide resource changes? They seem fine, although I haven't run the tests myself yet

AFAIK, the only difference with the subtests would be that if the memory part or cpu part fails, instead of saying "subtest failed: Container resources are correctly applied", it could say something more specific like `subtest failed: should be able to set set limits.cpu`, followed by a stacktrace. And of course, the fact that it acts as sort of substitute for the comments.